### PR TITLE
add square pcb plated hole

### PR DESCRIPTION
**Pull request by techmannih**
related to https://github.com/tscircuit/core/issues/429

### CODE REVIEW

**Comment by seveibar** [comment]
@techmannih i think it's ambiguous if the hole is square or the plating is square, circuit json should be unambiguous so we need to clarify

I'm not sure what @imrishabh18 was thinking about here, but it seems like we need to describe "square plating with circular hole"? Is that right @imrishabh18 ?

I would recommend picking a nice name for "shape", then also adding "hole_shape" and "plating_shape" for clarity

**Comment by seveibar** [comment]
Square holes are uncommon, so I think rishabh was likely talking about a circular hole with square plating as shown below

image

**Comment by imrishabh18** [comment]
Yeah, my fault I should have added the picture in the issue. I was talking about the square plate and circular hole

**Comment by seveibar** [comment]
this is the first time that we have an ambiguous situation, I'd recommend the shape name to be `shape: "circular_hole_with_square_pad"` and the clarifying `hole_shape: "circle"` and `pad_shape: "square"` to be added

it sounds like overkill but in it will be worth it when the specification is unambiguous!

**Review by seveibar — CHANGES_REQUESTED** [review]
reread my comments carefully, you didn't address them (not even close- you missed almost all of them?)

File: src/pcb/pcb_plated_hole.ts [hunk 1]
```diff
@@ -70,17 +70,58 @@ export interface PcbPlatedHoleOval {
   pcb_plated_hole_id: string
 }
 
+const pcb_circular_hole_with_square_plate = z.object({
+  type: z.literal("pcb_plated_hole"),
+  shape: z.literal("circular_hole_with_square_pad"),
```
Comment (seveibar) [inline src/pcb/pcb_plated_hole.ts:75]:
```suggestion
  shape: z.literal("circular_hole_with_rectangular_pad"),
```

File: src/pcb/pcb_plated_hole.ts [hunk 2]
```diff
@@ -70,17 +70,58 @@ export interface PcbPlatedHoleOval {
...
+  hole_shape: z.literal("circle"),
+  pad_shape: z.literal("square"),
+  hole_diameter: z.number(),
+  square_pad_width: z.number(),
+  square_pad_height: z.number(),
```
Comment (seveibar) [inline src/pcb/pcb_plated_hole.ts:80]:
```suggestion
  rect_pad_width: z.number(),
  rect_pad_height: z.number(),
```

File: src/pcb/pcb_plated_hole.ts [hunk 3]
```diff
@@ -70,17 +70,58 @@ export interface PcbPlatedHoleOval {
   pcb_plated_hole_id: string
 }
 
+const pcb_circular_hole_with_square_plate = z.object({
+  type: z.literal("pcb_plated_hole"),
+  shape: z.literal("circular_hole_with_square_pad"),
+  hole_shape: z.literal("circle"),
+  pad_shape: z.literal("square"),
```
Comment (seveibar) [inline src/pcb/pcb_plated_hole.ts:77]:
```suggestion
  pad_shape: z.literal("rect"),
```

File: src/pcb/pcb_plated_hole.ts [hunk 4]
```diff
@@ -70,17 +70,58 @@ export interface PcbPlatedHoleOval {
   pcb_plated_hole_id: string
 }
 
+const pcb_circular_hole_with_square_plate = z.object({
```
Comment (seveibar) [inline src/pcb/pcb_plated_hole.ts:73]:
```suggestion
const pcb_circular_hole_with_rect_pad = z.object({
```

File: src/pcb/pcb_plated_hole.ts [hunk 5]
```diff
@@ -70,17 +70,58 @@ export interface PcbPlatedHoleOval {
...
+  x: distance,
+  y: distance,
+  layers: z.array(layer_ref),
+  port_hints: z.array(z.string()).optional(),
+  pcb_component_id: z.string().optional(),
+  pcb_port_id: z.string().optional(),
+  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
+})
+
+export interface PcbHoleCircularWithSquarePlated {
```
Comment (seveibar) [inline src/pcb/pcb_plated_hole.ts:90]:
```suggestion
export interface PcbHoleCircularWithRectPad {
```

**Review by seveibar — CHANGES_REQUESTED** [review]
a square always has the same width/height, so it doesn't make sense to define a width and height unless we're doing a redudant definition, but we typically don't do this for rects (we usually just call it a rect and give it a width and height)

**Review by seveibar — APPROVED** [review]

**Pull request merged**